Adapt the `is_virtual` property value for Product Bundles to avoid applying incorrect shipping rates in products sync

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #1738 

This PR adapts the `is_virtual` property value for Product Bundles to avoid applying incorrect shipping rates in products sync

- Add a new filter to allow modifying product `is_virtual` property value in `WCProductAdapter`
- Adapt the `is_virtual` property for bundle products in `WooCommerceProductBundles` via the new filter
- Update src/Hooks/README.md

### Screenshots:

![2023-05-30 16 34 33](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/02ec29e0-d7ae-4b61-850d-e7449e5bf19a)

### Additional details:

Product Bundles extension uses another property to determine if a bundle product is virtual and set the original product virtual value to `yes` by default.

![2023-05-30 16 44 14](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/618ed426-997d-41ce-8106-131d58eae79c)

This causes the GLA to always apply shipping settings to bundle products:

https://github.com/woocommerce/google-listings-and-ads/blob/1f755140c690cee62170a71439433e47b9b9fab0/src/Product/WCProductAdapter.php#L448-L459

### Detailed test instructions:

💡 If testing in a local environment, it might need to bypass the mismatched URL check first:
- Add this code snippet to google-listings-and-ads.php for skipping the URL matching check.
   ```php
   add_filter( 'woocommerce_gla_ready_for_syncing', '__return_true' );
   ```
- Clear related transients from DB.
   ```sql
   DELETE FROM wp_options where option_name LIKE '%gla_url_matches%';
   ```

1. Install and activate [Product Bundles](https://github.com/woocommerce/woocommerce-product-bundles/releases) extension.
1. Create a new product with **Product bundle** type and some sub-products.
1. In GLA's free listings editing page, set "Shipping rates" to "My shipping settings are complex. I will enter my shipping rates and times manually in Google Merchant Center."
1. In Google Merchant Center, set different shipping settings.
1. Sync the bundle product (via the GLA Connection Test page).
1. Open the bundle product in Google Merchant Center.
1. Check if it shows "This shipping rate reflects the shipping rates set in the shipping settings." and the shipping rates are the same as the shipping settings in Google Merchant Center.
1. In GLA's free listings editing page, set "Shipping rates" to "My shipping settings are simple. I can manually estimate flat shipping rates."
1. Sync the bundle product (via the GLA Connection Test page).
1. Refresh the bundle product in Google Merchant Center.
1. Check if it still shows "This shipping rate reflects the shipping rates set in the shipping settings." and the shipping rates should be the same as the shipping settings in GLA.

💡 It usually takes a while for the Google Merchant Center to reflect products synchronization

### Changelog entry

> Fix - Adapt the `is_virtual` property value for Product Bundles to avoid applying incorrect shipping rates in products synchronization.

eason9487 committed:
Add a new filter to allow modifying product `is_virtual` property value in WCProductAdapter

eason9487 committed:
Adapt the `is_virtual` property for bundle products in WooCommerceProductBundles via filter

eason9487 committed:
Update src/Hooks/README.md

mikkamp commented on src/Integration/WooCommerceProductBundles.php:
```diff
@@ -189,6 +199,18 @@ private function calculate_sale_price( float $sale_price, WC_Product $product, b
+	 * @param WC_Product $product    WooCommerce product
+	 */
+	private function is_virtual_bundle( bool $is_virtual, WC_Product $product ): bool {
+		if ( $product instanceof WC_Product_Bundle && is_callable( [ new WC_Product_Bundle(), 'is_virtual_bundle' ] ) ) {
```
Why are we creating a new WC_Product_Bundle instance to check if the function is callable? We have already confirmed that `$product` is the right instance so we can use it in the `is_callable` check.

eason9487 replied:
Thanks! Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1967/commits/db6c57f6d2791bd4a83cd3a4e05429996f2fbfe9.

When I was looking for examples of `is_callable` usage, I was too focused on PHP 7/8 compatible syntax and missed there is already an existing `$product` instance.

mikkamp reviewed (approved):
I was having a bit of trouble replicating this scenario. If I go to create a new bundled product when I update the virtual setting it syncs both the meta `_virtual` and `_wc_pb_virtual_bundle`. Even if I manually unhide the checkbox and change it's value, when I update the product it stays in sync.

The only way I was able to replicate the scenario is by editing the product meta in the db and setting:
```
_virtual = yes
_wc_pb_virtual_bundle = no
```

In that scenario it overrides the shipping settings with 0 values, and when I sync the same product with this PR it's fixed because it gets the virtual setting directly from bundles.

However it makes me wonder why Bundles doesn't hook into the filter `woocommerce_is_virtual` and if the product is a bundle it checks `is_virtual_bundle`. Looking at the code it seems there might be cases where it could differ when products within the bundle are marked to ship individually. But I still can't find a way to replicate the difference in virtual setting. Which leaves me wondering if there are still certain configurations where it might not relay the correct shipping settings.

Either way I'll go ahead and approve the PR as it does allow the product bundle virtual setting to have priority which should make it easier to use the configured shipping methods.

eason9487 committed:
Use the existing instance to call `is_callable` instead of creating a new instance.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1967#discussion_r1210226509

eason9487 commented:
@mikkamp, thanks for the review.

> I was having a bit of trouble replicating this scenario. If I go to create a new bundled product when I update the virtual setting it syncs both the meta `_virtual` and `_wc_pb_virtual_bundle`. Even if I manually unhide the checkbox and change it's value, when I update the product it stays in sync.

I just created two new bundle products. Their product meta values are the following and the values won't be synced after I update the virtual setting via the product editing page.
```
_virtual = yes
_wc_pb_virtual_bundle = no
```

This might be due to some compatibility issues between Product Bundles and WooCommerce, and I happened to use versions that can reproduce this issue. The versions in my testing env:
- WP 6.0.5
- WC 7.6.0
- Product Bundles 6.18.6
